google-cloud-storage/src/main/java/com/google/cloud/storage/StorageImpl.java

```diff
@@ -733,6 +736,124 @@ public URL signUrl(BlobInfo blobInfo, long duration, TimeUnit unit, SignUrlOptio
     }
   }
 
+  @Override
+  public V4PostPolicy generateV4PresignedPostPolicy(
+      BlobInfo blobInfo,
+      V4PostFields fields,
+      V4PostConditions conditions,
+      long duration,
+      V4PostPolicyOption... options) {
+    EnumMap<SignUrlOption.Option, Object> optionMap = Maps.newEnumMap(SignUrlOption.Option.class);
+    // Convert to a map of SignUrlOptions so we can re-use some utility methods
+    for (V4PostPolicyOption option : options) {
+      optionMap.put(SignUrlOption.Option.valueOf(option.getOption().name()), option.getValue());
+    }
+
+    optionMap.put(SignUrlOption.Option.SIGNATURE_VERSION, SignUrlOption.SignatureVersion.V4);
+
+    ServiceAccountSigner credentials =
+        (ServiceAccountSigner) optionMap.get(SignUrlOption.Option.SERVICE_ACCOUNT_CRED);
+    if (credentials == null) {
+      checkState(
+          this.getOptions().getCredentials() instanceof ServiceAccountSigner,
+          "Signing key was not provided and could not be derived");
+      credentials = (ServiceAccountSigner) this.getOptions().getCredentials();
+    }
+
+    checkArgument(
+        !(optionMap.containsKey(SignUrlOption.Option.VIRTUAL_HOSTED_STYLE)
+            && optionMap.containsKey(SignUrlOption.Option.PATH_STYLE)
+            && optionMap.containsKey(SignUrlOption.Option.BUCKET_BOUND_HOST_NAME)),
+        "Only one of VIRTUAL_HOSTED_STYLE, PATH_STYLE, or BUCKET_BOUND_HOST_NAME SignUrlOptions can be"
+            + " specified.");
+
+    String bucketName = slashlessBucketNameFromBlobInfo(blobInfo);
+
+    boolean usePathStyle = shouldUsePathStyleForSignedUrl(optionMap);
+
+    String url =
```

Comment:
The ternary operator is hard to read when broken across multiple lines. Consider if else.

Reply:
Elliotte's comment here makes sense. Probably want to use an if-else in this case.